catkin_ws/src/controls/src/servers/state_quaternion_pid_server.py

```diff
@@ -14,36 +14,40 @@ def __init__(self):
         self.server = actionlib.SimpleActionServer('state_quaternion_server', StateQuaternionAction, execute_cb=self.callback, auto_start=False)
         self.previous_goal_quat = None
         # Calculation parameters/values
-        
+        self.goal_position = None
+        self.goal_quat = None
         self.server.start()        
 
     def callback(self, goal):
         print("\n\nQuaternion Server got goal:\n",goal)
         self.cancelled = False
-        self.goal = goal
         if self.pose is not None:
-            if(self.goal.displace.data):
-                goal_position, goal_quat = self.get_goal_after_displace()
+            if(goal.displace.data):
+                goal_position, goal_quat = self.get_goal_after_displace(goal)
+                self.goal_position = goal_position
+                self.goal_quat = goal_quat
             else:
-                goal_position = [self.goal.pose.position.x, self.goal.pose.position.y, self.goal.pose.position.z]
-                goal_quat = np.quaternion(self.goal.pose.orientation.w, self.goal.pose.orientation.x, self.goal.pose.orientation.y, self.goal.pose.orientation.z)
+                goal_position = [goal.pose.position.x, goal.pose.position.y, goal.pose.position.z]
+                goal_quat = np.quaternion(goal.pose.orientation.w, goal.pose.orientation.x, goal.pose.orientation.y, goal.pose.orientation.z)
+                self.goal_position = goal_position
```

Comment:
This is antoine from vasily's computer: this should be updated in the if statements, not for every goal. With this code, the goal positon/quaternion would always be updated to the most recent state goal, even if the do_x or d_y or dz or do_quat are false, in which case the goal position axes/quaternion should not be updated, since they will be 0 (instead of None). That's why I put the update for my implementation in the if goal.go_quat block so that it wouldn't be updated for position goals or wtv. Other than that LTGM